stan/math/rev/core/vari.hpp

```diff
@@ -3,13 +3,44 @@
 
 #include <stan/math/rev/core/chainable_alloc.hpp>
 #include <stan/math/rev/core/chainablestack.hpp>
+#include <stan/math/prim/meta.hpp>
 #include <ostream>
+#include <type_traits>
 
 namespace stan {
 namespace math {
 
 // forward declaration of var
-class var;
+template <typename T, typename>
```

Comment:
Why is there the second unnamed template argument?

Reply:
Oh is this something to do with these methods not being defined inside the class?

Reply:
This is so that var can be restricted to only floating point types via

```cpp
class var_value<T, require_floating_point_t<T>> {
```

Otherwise we had to do a weird reinterpret cast thing to deal with integral types

Reply:
I'm working on lining up the static matrix merge and seeing lots of stuff like:

```c++
template <typename T>
<<<<<<< HEAD                                                                                                                                                                 
template <typename Arith, require_arithmetic_t<Arith>...>                                                                                                                    
inline var_value<T>& var_value<T>::operator/=(const Arith& b) {                                                                                                              
=======                                                                                                                                                                      
template <typename Arith, require_arithmetic_t<Arith>*>                                                                                                                      
inline var_value<T, require_floating_point_t<T>>&                                                                                                                            
var_value<T, require_floating_point_t<T>>::operator/=(Arith b) {                                                                                                             
>>>>>>> origin/feature/vari-base-templates                                                                                                                                   
```

This seems related. Do you have merge advice?

Reply:
Oh I think I should be using the second form and replacing `require_floating_point_t` with `require_vt_floating_point`?

Reply:
Yeah I can update that branch with this later today if you don't want to slug through it

Reply:
Let me keep slugging for a bit. I might throw it all away and demand you update the branch, but until then it's probably a good education experience so that I can review this.